src/domain_decomposition.hpp

```diff
@@ -0,0 +1,153 @@
+#pragma once
+
+#include <vector>
+
+#include <backends.hpp>
+#include <common_types.hpp>
+#include <communication/global_policy.hpp>
+#include <recipe.hpp>
+#include <util/optional.hpp>
+#include <util/partition.hpp>
+#include <util/transform.hpp>
+
+namespace nest {
+namespace mc {
+
+// Meta data used to guide the domain_decomposition in distributing
+// and grouping cells.
+struct group_rules {
+    cell_size_type target_group_size;
+    backend_policy policy;
+};
+
+class domain_decomposition {
+    using gid_partition_type =
+        util::partition_range<std::vector<cell_gid_type>::const_iterator>;
+
+public:
+    /// Utility type for meta data for a local cell group.
+    struct group_range_type {
+        cell_gid_type begin;
+        cell_gid_type end;
+        cell_kind kind;
+    };
+
+    domain_decomposition(const recipe& rec, const group_rules& rules):
+        backend_policy_(rules.policy)
+    {
+        EXPECTS(rules.target_group_size>0);
+
+        auto num_domains = communication::global_policy::size();
+        auto domain_id = communication::global_policy::id();
+
+        // Partition the cells globally across the domains.
+        num_global_cells_ = rec.num_cells();
+        cell_begin_ = (cell_gid_type)(num_global_cells_*(domain_id/(double)num_domains));
+        cell_end_ = (cell_gid_type)(num_global_cells_*((domain_id+1)/(double)num_domains));
+
+        // Partition the local cells into cell groups that satisfy three
```

Comment:
I completely understand this now: thanx!

It even hints at possible refactorings